### ISSUE TITLE
term frequency and multi highlighting

### ISSUE THREAD

**Issue by leeoniya** [leeoniya/uFuzzy#19]
this might be more in the realm of a document search, but could be interesting to explore.

in info() or search() use a sticky regex to walk each matched haystack item with an altenation group, track frequency and have that somehow feed into sort(). also compute all highlight ranges.

also can establish some convention for haystack item parts, like heading|summary|content|footer, and ranges can be relative to partition points.

not sure how far this is worth taking. scope creep is a thing to keep in check.

**Comment by leeoniya** [comment]
`matchAll` could be useful for multiple out-of-order stuff, instead of sticky regex

![image](https://user-images.githubusercontent.com/43234/214711599-3cdf5a8c-84b1-4662-b0dd-fae35087117b.png)

**Comment by afonsopbarros** [comment]
is it already possible to have multi match highlighting using this internal interOr flag? @leeoniya

**Comment by leeoniya** [comment]
the internals rely on 1:1 mapping between needle terms and matched parts in the haystack items, and this is unlikely to change.

probably the best i can do is to expose the internal `u.prepQuery()` function and leave it up to the user to use `interOR` flag and run the resulting regexp via `.matchAll` to find all terms as they appear in any order in the located matches. this could potentially highlight quite unexpected portions of the matches that dont correspond to the actual ranking logic.

https://github.com/leeoniya/uFuzzy/blob/9486cf58cd937bc15082de217febfbf793bd589a/src/uFuzzy.js#L216C21-L216C21

**Comment by punkpeye** [comment]
so how do I highlight multiple matches within a string today?

`uf.search` just gives one range match

**Comment by punkpeye** [comment]
It is very confusing b/c the demo shows multiple highlights, but it seems like it is not possible to do with the current library?

**Comment by leeoniya** [comment]
> It is very confusing b/c the demo shows multiple highlights

this is the first time someone has found the highlighting behavior confusing.

if your needle is "my car", it will match and highlight "**my** fast **car**".

or are you refering to something different?

**Comment by punkpeye** [comment]
@leeoniya Just spent a solid hour on this until I realized that highlighting does not work as I would expect.

Haystack:

> foo bar foo

Needle:

> foo

I was expecting both foo instances to be highlighted.

**Comment by leeoniya** [comment]
yes, that's what this issue is about. but there are no demos that show this behavior because it is impossible for such demos to exist :)

sorry that you spent an hour before finding this.

**Comment by punkpeye** [comment]
It's alright. I just misunderstood the context of the issue and started with a bad test ase.

**Comment by punkpeye** [comment]
Even though it took a hecka of a time to figure out how to make this work (due to this minor hiccup and also figuring out how to search collections with multiple fields), the search results are a night and day difference compared to fuse.js.
